Core/Src/main.c

```diff
@@ -143,7 +158,43 @@ void SystemClock_Config(void)
 }
 
 /* USER CODE BEGIN 4 */
+void TIM1_CC_IRQHandler(void)
+{
+  // Interrupt handler code here
+  // reached desired pwm pulse width, so set PWM pin low
+  HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_RESET);
+}
+
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+  if (htim->Instance == TIM1)
+  {
+	HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8); // pull down Chip Select
+	uint8_t tx_data[3] = {0x01, 0x80, 0x00};
+	uint8_t rx_data[3] = {0};
+
+	HAL_SPI_TransmitReceive(&hspi1, tx_data, rx_data, ADC_SIZE, ADC_TIMEOUT);
+
+	// by anding the 2nd byte of rx_data with 0b0000011,
+	// we receive the 9th and 8th bit of the adc value
+	// then shifting them to the left by 8 bits to set them as the MSB
+	// oring the bits together to form the 10-bit ADC value
+	ADC_value = ((rx_data[1] & 0x03) << 8) | rx_data[2];
+
+	// the function gives the ARR value, the clock counts from 0 to ARR
+	// so add 1 to get the period of the clock
+	uint32_t TIM1_period = __HAL_TIM_GET_AUTORELOAD(&htim1) + 1;
 
+	// ADC_value / ADC_RANGE gives the duty cycle of the PWM signal
+	// Multiply it by the period gives the number of clock that the PWM signal should remain high
+	pwm_pulse_width = (ADC_value / ADC_RANGE) * TIM1_period;
```

Comment:
This calculation is slightly incorrect because the possible range of duty cycles is 0-100% but we want a duty cycle corresponding to an on time of 5-10ms which would not lead to a range of duty cycles between 0-100%...